### PR TITLE
HoC Initial Cached AI Client

### PR DESCRIPTION
**Pull request by tshaffercodeorg**
HoC should now default to fetching results from a cached response rather than querying OpenAI. Current selection is deterministic, behavior to be adjusted following subsequent ux tests. 

image

### CODE REVIEW

File: apps/tsconfig.json
Comment (breville) [inline apps/tsconfig.json:28]: Ah, is this so that we can do an `import` from `@cdo/static`?

File: apps/src/dance/DanceAiClient.ts
Comment (breville) [inline apps/src/dance/DanceAiClient.ts]: nit: we could make these filenames a bit more readable, e.g. `cached-palette-map.json`.  (We seem to be rather inconsistent in `static/` but kebab case seems relatively common.)

File: apps/src/dance/DanceAiClient.ts
Comment (breville) [inline apps/src/dance/DanceAiClient.ts]: nit: space after `//` and `.` at end of sentence.

File: apps/src/dance/DanceAiModal.tsx
Comment (breville) [inline apps/src/dance/DanceAiModal.tsx]: Is this intentional?  We should be using the emoji selected by the user.

File: apps/src/dance/DanceAiModal.tsx
Comment (breville) [inline apps/src/dance/DanceAiModal.tsx]: Shouldn't we be passing `value` to `doAi`?  We need to test that this code path still works.

File: apps/src/dance/types.ts
Comment (breville) [inline apps/src/dance/types.ts]: If these are only used in a single file, I'd actually recommend putting them in that file, and just using this file for types that are used in more than one source file.

File: apps/static/dance/ai/model/cachedbackgroundmap.json
Comment (breville) [inline apps/static/dance/ai/model/cachedbackgroundmap.json]: I suspect we will not be using these LLM explanations after all, in which case I think best to remove them for now, until we figure out exactly what we are going to do.

File: apps/tsconfig.json
Comment (bencodeorg) [reply apps/tsconfig.json:28]: Yes, although I saw after adding this you already had a way to do this. Seems like a positive to have import work with these static paths in TS though anyway)?

`const inputLibrary = require(`@cdo/static/dance/ai/${inputLibraryFilename}.json`);`

Comment (breville) [reply apps/tsconfig.json:28]: I suspect so.  @molly-moen or @sanchitmalhotra126 probably know for sure.

File: apps/src/dance/DanceAiModal.tsx
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiModal.tsx]: This is a dummy test, it does take in the actual value later and outputs the log. Probably a good idea to delete this for the final version regardless

File: apps/src/dance/DanceAiModal.tsx
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiModal.tsx]: Good catch, I'll make sure to pass in the value

File: apps/src/dance/types.ts
Comment (tshaffercodeorg) [reply apps/src/dance/types.ts]: Gotcha, migrated these over to the DanceAiClient file instead

File: apps/static/dance/ai/model/cachedbackgroundmap.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cachedbackgroundmap.json]: Noted, I'll keep them for now in the cached file just in case but will generate the google emoji json without them

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (jmkulwik) [inline apps/static/dance/ai/model/cached-spacy-background-map.json:179]: I don't see "higher power" or "angles" in this list - are they named something different?

Additionally, this list is 2 longer than the full list we see in the UI on the [standalone level](https://studio.code.org/s/dance-2019/lessons/1/levels/10). Are there some extras?

Tracking here: https://docs.google.com/spreadsheets/d/1-JG_pkFGLvYREU2dF6A3HTFswgrJGhR6xfJoWIsO1N4/edit#gid=0

File: apps/static/dance/ai/model/cached-spacy-palette-map.json
Comment (jmkulwik) [inline apps/static/dance/ai/model/cached-spacy-palette-map.json:57]: Looks like this list is missing "black and white" and "light" from the standalone level.

File: apps/static/dance/ai/model/cachedbackgroundmap.json
Comment (jmkulwik) [inline apps/static/dance/ai/model/cachedbackgroundmap.json]: Why are there two separate sets of json files?

File: apps/static/dance/ai/model/cached-spacy-palette-map.json
Comment (jmkulwik) [inline apps/static/dance/ai/model/cached-spacy-palette-map.json:1]: What does "spacy" indicate in this file name?

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (jmkulwik) [inline apps/static/dance/ai/model/cached-spacy-background-map.json:3]: Non-blocking comment for v2 (probably best to answer after merging). Can we use the emojis directly rather than our interpretation of the emojis?

File: apps/src/dance/DanceAiClient.ts
Comment (jmkulwik) [inline apps/src/dance/DanceAiClient.ts]: NIT: Please make a task for each TODO

File: apps/src/dance/DanceAiClient.ts
Comment (jmkulwik) [inline apps/src/dance/DanceAiClient.ts]: NIT: Do we still need this console log?

File: apps/src/dance/DanceAiClient.ts
Comment (jmkulwik) [inline apps/src/dance/DanceAiClient.ts]: Ask: rename this to an action-oriented name, such as "chooseEffects"
Reasoning: A good rule of thumb is to name functions with verbs ([useful cheat sheet I like](https://dev.to/maikomiyazaki/beginner-s-cheat-sheet-commonly-used-verbs-for-naming-functions-methods-and-variables-509i)). That way, it's clearer to the reader what action the function is carrying out. It also prevents the function from growing into doing something it shouldn't in the future. Having a noun here, "cachedAi" makes me think this is a variable and therefore will have the same value regardless of how I use it. Migrating to an action will clarify that confusion. This function also probably doesn't need the word "ai" in it since the file is named "DanceAiClient" and so it's known that this function is an action taken by an AI client (although I'm holding that opinion very loosely). Finally, I recommend against "cached" since that seems to be an implementation detail. Ideally, we'd have a "DanceAiClient" that would be created with the knowledge of whether it should use its cached version or query OpenAI directly. Then, later on, we'd call DanceAiClient.chooseEffects(emojis). DanceAiClient would have the freedom to act as an api client and query OpenAI or shortcut that process using its knowledge of a cache.

Comment (jmkulwik) [inline apps/src/dance/DanceAiClient.ts]: Separately, I recommend changing this so it is not a "default" export.

Reasoning: Given the filename DanceAiClient, I would expect this to be an API client, which it appears to be (but with the only action right now to be shortcutting the API call in favor of a cached version). With that, I would expect any default export (if there is one) to be a class. What would also seem normal would be to import the module and call available functions. For example, DanceAiClient.cachedAi(input). Importing this directly as DanceAiClient would make me expect that I get the whole client rather than a single function that returns the 3 specific effects.

Caveat: I'm not as familiar with TS as I am with vanilla JS, so if you know/find something that suggests a different convention, I'm happy to be proven wrong. I did find the [TS guide](https://www.typescriptlang.org/docs/handbook/modules.html), which seems to suggest (although not declare) that the above approach would be correct.

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (jmkulwik) [inline apps/static/dance/ai/model/cached-spacy-background-map.json:176]: I'm curious if a different json setup would simplify some of the code later. For example, something like this seems much easier to deal with:

```
{
  "emojiAssociations": {
    "cool": {
      "circles": 0.13,
      "color_cycle": 0.28,
      "diamonds": 0.11
    }
  }
}
```
I believe "emojiAssociations" also wouldn't be needed in this case, and it would decrease the amount of work the client would need to do. It would also make hand-checking (and manual adjustments) much easier.

File: apps/src/dance/DanceAiClient.ts
Comment (jmkulwik) [inline apps/src/dance/DanceAiClient.ts]: This seems more complicated than it needs to be. Would a simple loop that iterates over the cached backgrounds, foregrounds, etc work?

Reasoning: This feels a bit esoteric. When I see Object.prototype.hasOwnProperty, it makes me wonder if there's some code smell here. I don't see this sequence used very often, and I definitely had to google it. I get the feeling I'm not the only one since the [mdn docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) have the line "Here be dragons" as they explain it 😅. 

I could see something like this working:
```
chosenEmojis = [cool, magic, wavy]
foregroundMap = foregroundValues.json
// if the json is reformatted as I recommend in a later comment, the next step isn't needed
reformattedForegroundValues = {}
Object.keys(foregroundMap.emojiAssociations).forEach((emoji) => {
  reformattedForegroundValues[emoji] = {}
  foregroundMap.emojiAssociations[emoji].association.forEach((calculatedValue, index) => {
      reformattedForegroundValues[emoji][foregroundMap.output[index]] = calculatedValue
  })
})
// ^^ then same for reformatting the other files
// end code that could be skipped

// Here, we have a few options. 

// If we want a list of values so we can choose a weighted random value:
var summedForegroundValues = [];
var foregroundEffects = [];
Object.keys(reformattedForegroundValues[chosenEmojis[0]]).forEach((effect) => {
    summedForegroundValues2.push(reformattedForegroundValues[chosenEmojis[0]][effect] + reformattedForegroundValues[chosenEmojis[1]][effect] + reformattedForegroundValues[chosenEmojis[2]][effect])
    foregroundEffects.push(effect)
})

// Or, if we want a mapping of emoji to value
var summedForegroundValues2 = {}
Object.keys(reformattedForegroundValues[chosenEmojis[0]]).forEach((effect) => {
  summedForegroundValues[effect] = reformattedForegroundValues[chosenEmojis[0]][effect] + reformattedForegroundValues[chosenEmojis[1]][effect] + reformattedForegroundValues[chosenEmojis[2]][effect]
})

// Or, if we just want the highest value
var highestValue = Object.keys(reformattedForegroundValues[chosenEmojis[0]]).reduce((highestValue, effect) => {
  return Math.max(highestValue, reformattedForegroundValues[chosenEmojis[0]][effect] + reformattedForegroundValues[chosenEmojis[1]][effect] + reformattedForegroundValues[chosenEmojis[2]][effect])
}, 0)

```

File: apps/static/dance/ai/model/cachedbackgroundmap.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cachedbackgroundmap.json]: These were older JSONs for our in-house emojis while we were testing with them; we can delete them now if we don't want them anymore since we don't actively use them

File: apps/static/dance/ai/model/cached-spacy-palette-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-palette-map.json:1]: SpaCy is the python library/model we used to generate these so I named them that although I totally get why that might be confusing

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Nope, I'll delete it

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Definitely agree with all of this, the function was named cachedAi to distinguish the source from the prior function that queryed ChatGPT named "doAi". We can just switch the naming conventions of both to be more indicative of what's actually going on though

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-background-map.json:176]: emojiAssociations existing as a separate nested json is due to typescript not playing well with mixed types in one json. It also mattered more when there was a pre-generated explanation sentence with each entry; definitely agree that there might be a more elegant solution now that's not an issue.

However, the current list format of encoded values was chosen to make summing three different vectors easier so I'm not sure if iterating through each key/value pairing necessarily makes that process any easier

File: apps/src/dance/DanceAiClient.ts
Comment (bencodeorg) [reply apps/src/dance/DanceAiClient.ts]: I think I created this file before we created a utils file and moved the existing `doAi` there:

https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/dance/ai/utils.ts

Might make sense to move this function into utils as well?

Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Let's review this as a next step but I generally agree that it seems strange to import a whole class just for one function, even if that class has a fair number of dependencies; I'm also unfamiliar with typescript protocols in that regard

Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: How about we make migrating this into utils as a next-step task to help unblock other folks?

File: apps/static/dance/ai/model/cached-spacy-palette-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-palette-map.json:57]: Those were deleted due to the current HoC not having render options for those two effects

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-background-map.json:179]: Those were deleted due to the current HoC not having render options for those two effects, similar to the item below. Because there is currently parallel work on re-integrating those two items right now, I'd suggest making bringing those missing options back in as a next-steps item

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Should this be in JIRA/Trello?

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: It took me a bit to get where you coming from but I think I have a formatting solution for this; I think originally we wanted to be able to separately access the vector of values and any generated text that came with it. However, now that the JSONs are JUST vectors of values, we can sum them all directly.

Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Documenting this follow-up task here https://trello.com/c/Xt0uC2ai

File: apps/tsconfig.json
Comment (tshaffercodeorg) [reply apps/tsconfig.json:28]: Is there any follow-up action item to this?

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Will be performing a larger reformatting to simplify substeps as described in this trello task: https://trello.com/c/Xt0uC2ai

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Tracking this as a next step task in this Trello ticket: https://trello.com/c/KvfZFqlX

Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: Consolidating this with the trello ticket task above to generally just migrate this into the utils file: https://trello.com/c/KvfZFqlX

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-background-map.json:176]: Consolidating this with the comment below as part of a broader eval of reformatting our code now that we no longer expect specific data types/structures to be supported: https://trello.com/c/Xt0uC2ai

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-background-map.json:3]: See follow-up discussion on this point within this slack thread: https://codedotorg.slack.com/archives/C05P12U42F5/p1697067515602939

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-background-map.json:179]: Tracking this follow-up in this trello ticket here: https://trello.com/c/856OLXRA

File: apps/static/dance/ai/model/cached-spacy-palette-map.json
Comment (tshaffercodeorg) [reply apps/static/dance/ai/model/cached-spacy-palette-map.json:57]: Similar to above comment, tracking regenerating to support these missing options later here: https://trello.com/c/856OLXRA

File: apps/tsconfig.json
Comment (bencodeorg) [reply apps/tsconfig.json:28]: I don't think so, think it was more of a clarification

File: apps/src/dance/DanceAiClient.ts
Comment (jmkulwik) [inline apps/src/dance/DanceAiClient.ts]: I noticed that these caused lint errors when I tried to commit. Please fix all lint errors before merging.

File: apps/src/dance/DanceAiClient.ts
Comment (jmkulwik) [inline apps/src/dance/DanceAiClient.ts]: I was playing around with this and noticed that this function only returns the highest value of the second emoji. The first and third are ignored. Or, put a different way, it runs the correct calculations, but only as if the second emoji is all that was input by the user. I made some notes about this in a separate branch: https://github.com/code-dot-org/code-dot-org/pull/54217. I also took the opportunity to see if a different algorithm could produce correct results for the combination of all three emojis. The results of this investigation are also in the linked branch.

**Review by jmkulwik — CHANGES_REQUESTED** [review]
There were two specific issues I noticed when I pulled the branch:

1. There are a few lint errors. These will cause errors for anyone who merges staging once this has gone in, so they should be addressed before merging.
2. The algorithm as-written does not calculate the values for the 1st and 3rd emoji - only the 2nd. From what I can tell, the 1st and 3rd are ignored. I made some notes in a separate branch as to why that might be https://github.com/code-dot-org/code-dot-org/pull/54217. I also tried out a different algorithm to see if it could produce correct results for the combination of all three emojis. The results of that investigation are also in the linked branch.

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: I found out what was causing the error; there was a naive assumption that the input would strictly follow the format "[emoji1], [emoji2], [emoji3]" but the actual input was structured more like a naturalistic sentence. Therefore, when it would search the keys:
1 would not return due to having the phrase "Return a blah blah blah [emoji1]" and
3 would not return due to having an addition period at the end e.g. "silly." vs "silly"

The program still ran since it would ONLY attempt to pull a value if the JSON contained that key; a great security fail-safe but definitely something that caused me to miss that during testing. Upon removing that check, the algorithm immediately crashed from the above.

Inputting JUST the emojiNames as a list in the expected format causes both algorithms to work.

In the interim, I'll merge your changes as described and see if we can incorporate some of the suggested JSON reformatting into the generation script itself rather than having the native file do this reformatting in 2.0

File: apps/src/dance/DanceAiClient.ts
Comment (tshaffercodeorg) [reply apps/src/dance/DanceAiClient.ts]: That's super weird, maybe it's because of the any? I have VSCode auto-run the prettier lint whenever I save so I'm surprised it didn't catch that

**Review by jmkulwik — COMMENTED** [review]
Sorry, realized my change request was blocking this PR. Removing that!

**Review by jmkulwik — APPROVED** [review]

File: apps/src/dance/DanceAiClient.ts
Comment (jmkulwik) [reply apps/src/dance/DanceAiClient.ts]: (checking in from the airport)
Ahh that makes sense. I completely missed the `.` when I was looking at it. When I was thinking about it again this morning, I think I overdid the reformatting. I'd love to chat more on Monday about the best structure for the JSON file. 

I'm fine with merging as is now that we have an accurate algorithm. There's two changes I'd request however. First, I'd love to maintain both DanceAiClient and DanceAiClient2 so we can cross-reference the algorithms as we continue to refine the code here. However, they shouldn't both exist in production. So long as you don't delete your local branch (and/or I don't delete mine), they'll continue to exist in the git history. To merge this Pull Request, however, one should be removed in the PR (again, it'll be maintained in our local git histories, so we'll still be able to access it). As part of that, `DanceAiClient2` should be renamed to `DanceAiClient`. Additionally, there are a few comments that disable some code in `DanceAiModal` such as line 14 and line 139 that should be cleaned up.

File: apps/tsconfig.json
Comment (sanchitmalhotra126) [reply apps/tsconfig.json:28]: I'd say `import` is preferred if it works - using `require` allows us to circumvent TypeScript path aliasing which is helpful for certain files with specific knowns aliases (like @cdo/locale) but it removes any type checking and is basically an escape hatch to defer module resolution to webpack at runtime.

File: apps/static/dance/ai/model/cached-spacy-background-map.json
Comment (sanchitmalhotra126) [inline apps/static/dance/ai/model/cached-spacy-background-map.json:1]: For my understanding - I figured we'd generate and store these JSON files in S3 so that we could iterate on them without requiring a deploy. Has the thinking around that changed?

Comment (breville) [reply apps/static/dance/ai/model/cached-spacy-background-map.json:1]: Great question.  When we learned more about the incoming model, it turned out we didn't need to cache a combinatorial explosion of options on the server, because we are actually caching three relatively small sets of data and do the final combination on the client.  We were planning on snapshotting these three cache files in the repo anyway, and when it looked like the format of that data would be fairly tightly coupled to the client-side code, we decided to just rely on the local files.  We can certainly add an optional retrieval from S3 later, if it helps accelerate experimentation.

**Pull request merged**
